### PR TITLE
Add parameterizations and sensitivities to `SolidFunctional`

### PR DESCRIPTION
**Pull request by jamiebramwell**
This adds arbitrary parameterizations of load and materials to the `SolidFunctional` physics model. Included are two parameterized models: linear elastic and neo-Hookean responses with user-parameterized bulk and shear modulus. The interface could easily be extended to include both new material models as well as new parameterizations ready for sensitivity analysis.

### CODE REVIEW

File: src/serac/physics/tests/serac_solid_functional.cpp [hunk 1]
```diff
@@ -278,8 +277,99 @@ void functional_solid_test_boundary(double expected_disp_norm, TestType test_mod
...
+      StateManager::newState(FiniteElementState::Options{.order = 1, .name = "parameterized_shear"}));
+
+  user_defined_shear_modulus = 1.0;
+
+  FiniteElementState user_defined_bulk_modulus(
+      StateManager::newState(FiniteElementState::Options{.order = 1, .name = "parameterized_bulk"}));
+
+  user_defined_bulk_modulus = 1.0;
+
+  // Construct a functional-based thermal conduction solver
```
Comment (barreracruz) [inline src/serac/physics/tests/serac_solid_functional.cpp]: structural/nonlinear mechanics solver?

File: src/serac/physics/tests/serac_solid_functional.cpp [hunk 2]
```diff
@@ -278,8 +277,99 @@ void functional_solid_test_boundary(double expected_disp_norm, TestType test_mod
...
+  SolidFunctional<p, dim, H1<1>, H1<1>> solid_solver(default_static, GeometricNonlinearities::On,
+                                                     FinalMeshOption::Reference, "solid_functional",
+                                                     {user_defined_bulk_modulus, user_defined_shear_modulus});
+
+  solid_util::ParameterizedNeoHookeanSolid<dim> mat(1.0, 0.0, 0.0);
+  solid_solver.setMaterial(mat);
+
+  // Define the function for the initial temperature and boundary condition
```
Comment (barreracruz) [inline src/serac/physics/tests/serac_solid_functional.cpp]: initial displacement

File: src/serac/physics/tests/serac_solid_functional.cpp [hunk 3]
```diff
@@ -278,8 +277,99 @@ void functional_solid_test_boundary(double expected_disp_norm, TestType test_mod
...
+  auto bc = [](const mfem::Vector&, mfem::Vector& bc_vec) -> void { bc_vec = 0.0; };
+
+  // Set the initial temperature and boundary condition
```
Comment (barreracruz) [inline src/serac/physics/tests/serac_solid_functional.cpp]: Ditto

File: src/serac/physics/tests/serac_solid_functional.cpp (hunk 1 above)
Comment (jamiebramwell) [reply src/serac/physics/tests/serac_solid_functional.cpp]: Arg! My copy/paste leavings! I thought I got all of them but I'll do another pass.

File: src/serac/physics/solid_functional.hpp [hunk 4]
```diff
@@ -24,6 +24,38 @@
...
+  NonlinearSolverOptions H_nonlin_options;
+
+  /**
+   * @brief The optional ODE solver parameters
+   * @note If this is not defined, a quasi-static solve is performed
+   */
+  std::optional<TimesteppingOptions> dyn_options = std::nullopt;
+};
+}  // namespace solid_util
+
```
Comment (samuelpmishLLNL) [inline src/serac/physics/solid_functional.hpp:58]: these appear to be effectively identical types to https://github.com/LLNL/serac/blob/develop/src/serac/physics/thermal_conduction_functional.hpp#L29-L57, except that they can't be used interchangeably because they're declared in separate namespaces.

Do these options have anything to do with the specific physics? If not, I think there should be a single definition, outside the physics module namespaces, and be shared by all of the different modules.

File: src/serac/physics/materials/functional_material_utils.hpp [hunk 5]
```diff
@@ -198,50 +188,57 @@ auto parameterizeFlux(T& flux)
...
+ * @brief Wrapper to treat an unparameterized source as a parameterized one
+ *
+ * @tparam UnparameterizedSourceType The unparameterized source
+ */
+template <typename UnparameterizedPressureType>
+struct TriviallyParameterizedPressure {
+  /**
+   * @brief Wrapper for an unparameterized source in a parameterized context
+   *
+   * @tparam T1 Spatial position type
```
Comment (goxberry) [inline src/serac/physics/materials/functional_material_utils.hpp]: If `T2` and `T3` aren't arguments to this template, is it necessary to document them?

File: src/serac/physics/materials/parameterized_solid_functional_material.hpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                                    double bulk_modulus_offset = 1.0)
+      : density_(density), bulk_modulus_offset_(bulk_modulus_offset), shear_modulus_offset_(shear_modulus_offset)
+  {
+    SLIC_ERROR_ROOT_IF(density_ < 0.0, "Density must be positive in the linear isotropic elasticity material model.");
+  }
+
+  /**
+   * @brief Material response call for a linear isotropic solid
+   *
+   * @tparam T1 Spatial position type
```
Comment (goxberry) [inline src/serac/physics/materials/parameterized_solid_functional_material.hpp]: As someone potentially using this API, I'd find it easier to understand and use if the `T1` through `T5` parameters were replaced with, e.g., `SpatialPositionType`, particularly if the intent is for users to emulate this pattern.

File: src/serac/physics/materials/solid_functional_material.hpp [hunk 7]
```diff
@@ -52,30 +76,24 @@ class LinearIsotropicSolid {
...
-   * @brief Function defining the Kirchoff stress (constitutive response)
+   * @brief Material response call for a linear isotropic solid
    *
-   * @tparam T type of the Kirchoff stress (i.e. second order tensor)
+   * @tparam T1 Spatial position type
+   * @tparam T2 Displacement type
+   * @tparam T3 Displacement gradient type
    * @param du_dX displacement gradient with respect to the reference configuration (du_dX)
-   * @return The Kirchoff stress (det(deformation gradient) * Cauchy stress) for the constitutive model
+   * @return The calculated material response (density, kirchoff stress) for the material
```
Comment (goxberry) [inline src/serac/physics/materials/solid_functional_material.hpp]: kirchoff -> Kirchhoff

File: src/serac/physics/materials/solid_functional_material.hpp [hunk 8]
```diff
@@ -119,14 +137,16 @@ class NeoHookeanSolid {
...
-   * @brief Function defining the Kirchoff stress (constitutive response)
+   * @brief Material response call for a neo-Hookean solid
    *
-   * @tparam T type of the Kirchoff stress (i.e. second order tensor)
+   * @tparam T1 Spatial position type
+   * @tparam T2 Displacement type
+   * @tparam T3 Displacement gradient type
    * @param du_dX displacement gradient with respect to the reference configuration (du_dX)
-   * @return The Kirchoff stress (det(deformation gradient) * Cauchy stress) for the constitutive model
+   * @return The calculated material response (density, kirchoff stress) for the material
```
Comment (goxberry) [inline src/serac/physics/materials/solid_functional_material.hpp]: See https://github.com/LLNL/serac/pull/658/files#r841976690.

**Comment by goxberry** [comment]
AFAICT, LGTM.

**Review by goxberry — APPROVED** [review]

File: src/serac/physics/solid_functional.hpp (hunk 4 above)
Comment (jamiebramwell) [reply src/serac/physics/solid_functional.hpp:58]: This will be resolved in a follow on PR resolving https://github.com/LLNL/serac/issues/653 .

File: src/serac/physics/materials/functional_material_utils.hpp (hunk 5 above)
Comment (jamiebramwell) [reply src/serac/physics/materials/functional_material_utils.hpp]: Good point. I'm surprised our doxygen bot didn't notice this.

File: src/serac/physics/materials/parameterized_solid_functional_material.hpp (hunk 6 above)
Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: I'm OK with either. Do you have an opinion @samuelpmishLLNL @btalamini @white238?

Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: I just think this convention should be consistent. I followed the `T` style after tensor/functional, but the types here definitely have a physical meaning.

Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: I remember @tepperly having a similar question, so I think two potential users bring this up means we should change it.

Comment (samuelpmishLLNL) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: I think it does make sense to use more descriptive template parameter names here, since these quantities aren't so abstract. For some of the lower-level types (e.g. `tensor`, `tuple`, etc) those template parameters can represent practically anything (so it's harder to attach a descriptive name), whereas these ones are much more specific.

Also, we only need to use a template parameter for arguments that might be promoted to AD-enabled versions, so I think the spatial position could just be `tensor< double, dim > x`, rather than a template parameter `SpatialPositionType x`.

The naming of arguments here is also a little inconsistent, we're using `x`, `du_dX` (mathematical notation) for some arguments, and then `displacement`, `bulk_parameter`, `shear_parameter` (descriptive words) for the others. I'm not sure which is better, maybe a compromise like:

```cpp
  template <typename displacement_type, 
            typename displacement_gradient_type, 
            typename bulk_modulus_type, 
            typename shear_modulus_type >
   SERAC_HOST_DEVICE auto operator()(const tensor< double, dim > & /* x */, 
                                     const displacement_type & /* u */, 
                                     const displacement_gradient_type & du_dX, 
                                     const bulk_modulus_type & K, 
                                     const shear_modulus_type & G) const { ... }
```

File: src/serac/physics/materials/parameterized_solid_functional_material.hpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    auto shear_modulus = shear_parameter + shear_modulus_offset_;
+
+    auto I      = Identity<dim>();
+    auto lambda = bulk_modulus - (2.0 / dim) * shear_modulus;
+    auto strain = 0.5 * (du_dX + transpose(du_dX));
+    auto stress = lambda * tr(strain) * I + 2.0 * shear_modulus * strain;
+
+    using StressType = decltype(stress);
+
+    return MaterialResponse<double, StressType>{.density = density_, .stress = stress};
```
Comment (samuelpmishLLNL) [inline src/serac/physics/materials/parameterized_solid_functional_material.hpp]: The `decltype` here looks unpleasant, does
```cpp
return MaterialResponse{.density = density_, .stress = stress};
```
or
```cpp
return MaterialResponse{density_, stress};
```
work just the same? 

If not, I think we might be able to change the implementation of `MaterialResponse` type-- I really don't want the user to be required to manually deduce certain types in their q-functions.

Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: Let me check - I had to do a lot of compiler error chasing to get this to work. I'll see if it was something else I needed.

Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: If I recall correctly, clang couldn't resolve the class template argument deduction properly. I'll test this though.

Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: Well this seems to be working now, so I'll remove the explicit template types.

File: src/serac/physics/materials/parameterized_solid_functional_material.hpp (hunk 6 above)
Comment (goxberry) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: > For some of the lower-level types (e.g. tensor, tuple, etc) those template parameters can represent practically anything (so it's harder to attach a descriptive name)

I agree. I'm not terribly concerned about terse template arguments for low-level, algebraic data types; I'm mostly concerned with arguments that have physical significance.

re: `x`, `u` versus `position`, `displacement`, as someone without a solid mechanics background, I find the latter more helpful due to collisions in notation namespaces for different disciplines (e.g., `mu` could be a shear modulus, chemical potential, coefficient of static friction, or equality-constraint duality multiplier, depending on context). The longer names are also easier to `grep` when making changes, easier to Google if I'm unfamiliar with the physics, and less likely to become stale than explanatory comments.

File: src/serac/physics/materials/parameterized_solid_functional_material.hpp (hunk 9 above)
Comment (jamiebramwell) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: I spoke too soon! The linear version works, but the neo-Hookean model doesn't. I will investigate.

Comment (samuelpmishLLNL) [reply src/serac/physics/materials/parameterized_solid_functional_material.hpp]: if something like
```cpp
return MaterialResponse{.density = density_, .stress = stress};
```
doesn't work, I don't think it should hold up this PR. We can just put a `// TODO` on it and make an issue, to revisit later.

**Comment by jamiebramwell** [comment]
@samuelpmishLLNL @btalamini any other issues in this PR?

**Comment by jamiebramwell** [comment]
@samuelpmishLLNL @sethwatts @btalamini any other comments on this PR?

**Review by samuelpmishLLNL — APPROVED** [review]

**Comment by jamiebramwell** [comment]
ARG! GCC 11 got me! 😆 I'll investigate.
